turbot/resource_turbot_resource.go

```diff
@@ -100,75 +113,135 @@ func resourceTurbotResourceCreate(d *schema.ResourceData, meta interface{}) erro
-		var err error = nil
-		properties, err = helpers.PropertyMapFromJson(d.Get("data").(string))
-		if err != nil {
-			return fmt.Errorf("error retrieving properties from resource data: %s", err.Error())
-		}
-	}
-
-	resource, err := client.ReadResource(id, properties)
+	var err error
+	// read full resource
+	resource, err := client.ReadFullResource(id)
 	if err != nil {
 		if apiClient.NotFoundError(err) {
 			// resource was not found - clear id
 			d.SetId("")
 		}
 		return err
 	}
+	// build required properties from schema for data and metadata.
+	// properties is a map of property name -> property path
+	var dataProperties map[string]string
+	var metadataProperties map[string]string
 
-	// rebuild data from the resource
-	data, err := helpers.MapToJsonString(resource.Data)
-	if err != nil {
-		return fmt.Errorf("error building resource data: %s", err.Error())
+	// rebuild data, metadata from resource,
+	// include the properties that are in resource schema
+	if _, ok := d.GetOk("data"); ok {
+		dataProperties, err = getPropertiesFromResourceSchema(d, "data")
+		if err != nil {
+			return err
+		}
+		data := buildResourceMapFromProperties(resource.Data, dataProperties)
+		dataString, err := helpers.MapToJsonString(data)
+		if err != nil {
+			return fmt.Errorf("error building resource data: %s", err.Error())
+		}
+		d.Set("data", dataString)
+	}
+	if _, ok := d.GetOk("metadata"); ok {
+		metadataProperties, err = getPropertiesFromResourceSchema(d, "metadata")
+		if err != nil {
+			return err
+		}
+		metadata := buildResourceMapFromProperties(resource.Turbot.Custom, metadataProperties)
+		metadataString, err := helpers.MapToJsonString(metadata)
+		if err != nil {
+			return fmt.Errorf("error building resource data: %s", err.Error())
+		}
+		d.Set("metadata", metadataString)
 	}
 
-	// assign results back into ResourceData
+	// full_data and full_metadata attributes are directly read from readFullResource response
+	// we don't exclude any keys from the read response.
+	if _, ok := d.GetOk("full_metadata"); ok {
+		if metadata, err := helpers.MapToJsonString(resource.Turbot.Custom); err != nil {
+			d.Set("full_metadata", metadata)
+		}
+	}
+	if _, ok := d.GetOk("full_data"); ok {
+		if data, err := helpers.MapToJsonString(resource.Data); err != nil {
+			d.Set("full_data", data)
+		}
+	}
 
 	// set parent_akas property by loading resource and fetching the akas
 	if err := storeAkas(resource.Turbot.ParentId, "parent_akas", d, meta); err != nil {
 		return err
 	}
-
+	// assign results back into ResourceData
 	d.Set("parent", resource.Turbot.ParentId)
 	d.Set("type", resource.Type.Uri)
 	d.Set("tags", resource.Turbot.Tags)
-	d.Set("data", data)
 	return nil
 }
 
 func resourceTurbotResourceUpdate(d *schema.ResourceData, meta interface{}) error {
 	client := meta.(*apiClient.Client)
 	// build input map to pass to mutation
 	id := d.Id()
+	// build input data mutations parsing a list of top level properties permitted for update
```

Comment:
// build mutation input data by parsing the resource schema and excluding top level properties which must not be sent to update